`derive(FromBytes)` should implicitly derive `FromZeros` and `TryFromBytes`

Should deriving `FromBytes` imply deriving `FromZeros` and `TryFromBytes`? Either possible decision on this question has distinct advantages, disadvantages, and precedence. After reviewing the below design document and soliciting feedback from key customers, we've tentatively concluded *yes*.

## Background
Zerocopy defines the following trait hierarchy:
```rust
unsafe trait TryFromBytes { ... }
unsafe trait FromZeros: TryFromBytes { ... }
unsafe trait FromBytes: FromZeros { ... }
```

Implementing these traits manually is an explicitly unsupported workflow; they must be derived.

## Question
To derive `FromBytes`, should a user write:
- ...this:
    ```rust
    #[derive(TryFromBytes, FromZeros, FromBytes)]
    struct Foo {
        ...
    }
    ```
- ...or this:
    ```rust
    #[derive(FromBytes)]
    struct Foo {
        ...
    }
    ```

That is, *should deriving a zerocopy trait implicitly derive its super traits?*

## Evaluation

### Pedagogy
How does this decision impact our documentation? The zerocopy documentation makes heavy use of examples to demonstrate its constructs.

#### Option 1
Under Option 1, our examples must include derives for the full set of super traits, even when those traits are not germane to the example.

This has the unfortunate effect of making simple examples appear more complicated than they otherwise would be. However, the repeated exposure to the trait hierarchy this provides may reinforce the hierarchy in users' minds.

#### Option 2
Under Option 2, our examples need only include derives for traits germane to the example. This simplifies examples, at the expense of reducing the visibility of zerocopy's trait hierarchy. 

### Pathological User Experience
How does this decision impact our users? We'll consider the stories of Asher and Willow. Asher is using a version of zerocopy that does *not* imply the derives, but their user experience would likely be smoother if it did. Willow is using a version of zerocopy that *does* imply the derives, but their user experience would likely be smoother if it did not.

#### Option 1
Asher needs to call an API that requires its parameter implement `FromBytes`. Dutifully, they `derive` the trait:
```rust
#[derive(FromBytes)]
struct Foo {
    ...
}
```
...and compile. They receive the error message:
```plaintext
error[E0277]: the trait bound `Foo: FromZeroes` is not satisfied
    --> src/lib.rs:3:10
     |
3    | #[derive(FromBytes)]
     |          ^^^^^^^^^ the trait `FromZeroes` is not implemented for `Foo`
     |
     = help: the following other types implement trait `FromZeroes`:
               bool
               char
               isize
               i8
               i16
               i32
               i64
               i128
             and 67 others
note: required by a bound in `zerocopy::FromBytes`
    --> /playground/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zerocopy-0.7.32/src/lib.rs:1803:29
     |
1803 | pub unsafe trait FromBytes: FromZeroes {
     |                             ^^^^^^^^^^ required by this bound in `FromBytes`
     = note: this error originates in the derive macro `FromBytes` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
```
...and infer that they need to *also* implement `FromZeros`. So they add that to the `derive(...)`:
```rust
#[derive(FromZeros, FromBytes)]
struct Foo {
    ...
}
```
...and compile. They receive the error message:
```plaintext
error[E0277]: the trait bound `Foo: TryFromBytes` is not satisfied
    --> src/lib.rs:3:10
     |
3    | #[derive(FromZeros, FromBytes)]
     |          ^^^^^^^^^ the trait `TryFromBytes` is not implemented for `Foo`
     |
     = help: the following other types implement trait `TryFromBytes`:
               bool
               char
               isize
               i8
               i16
               i32
               i64
               i128
             and 67 others
note: required by a bound in `zerocopy::FromZeros`
    --> /playground/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zerocopy-0.7.32/src/lib.rs:1803:29
     |
1803 | pub unsafe trait FromZeros: TryFromBytes {
     |                             ^^^^^^^^^^^^ required by this bound in `FromZeros`
     = note: this error originates in the derive macro `FromBytes` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
```
...and infer that they need to *also* implement `TryFromBytes`. So they add that to the `derive(...)`:
```rust
#[derive(TryFromBytes, FromZeros, FromBytes)]
struct Foo {
    ...
}
```
...and compile. This compilation succeeds without errors.

#### Option 2
Willow has already derived `FromBytes` on their type:
```rust
#[derive(FromBytes)]
struct Foo {
    ...
}
```
They want to pass their `Foo` to a function that requires a `FromZeros` bound. They add the function call, but before trying to compile (which would succeed), they infer that they also need to derive `FromZeros`:
```rust
#[derive(FromZeros, FromBytes)]
struct Foo {
    ...
}
```
When they compile, they encounter an error message:
```plaintext
error[E0119]: conflicting implementations of trait `zerocopy::FromZeros` for type `Foo`
 --> src/lib.rs:3:22
  |
3 | #[derive(FromZeros, FromBytes)]
  |          ---------  ^^^^^^^^^ conflicting implementation for `Foo`
  |          |
  |          first implementation here
  |
  = note: this error originates in the derive macro `FromZeros` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0119`.
```
They delete their added derive of `FromZeros` and then compile without errors.

### Precedent
What precedent does either option have that might impact the expectations of our users?

#### Option 1
In the Rust standard libary, there is no direct analogy for the situation we are considering. At first glance, the `Copy: Clone` hierarchy in the Rust seems analogous. If Rust *could* have `#[derive(Copy)]` imply `#[derive(Copy, Clone)]`, then why doesn't it?

The premise, unfortunately, does not hold. If deriving `Copy` implied deriving `Clone`, then `Clone` could not be manually implementable:
```rust
#[derive(Clone)]
struct Foo {
    ...
}

impl Clone for Foo {
    fn clone(&self) -> Self { 
        todo!("a bespoke clone implementation")
    }
}
```
Compiling this would (in our hypothetical alternate `derive(Copy)`) produce an error:
```plaintext
error[E0119]: conflicting implementations of trait `Clone` for type `Foo`
 --> src/lib.rs:1:10
  |
1 | #[derive(Copy)]
  |          ^^^^ conflicting implementation for `Foo`
...
6 | impl Clone for Foo {
  | ------------------ first implementation here
  |
  = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0119`.
```
An [RFC proposing exactly this sort of mechanism](https://github.com/Centril/rfcs/blob/rfc/implied-derive/text/0000-implied-derive.md) for standard library traits was postponed because of exactly these sorts of backwards compatibility challenges.

A difference between this and our scenario is that our traits are *not* manually implementable; the backwards compatibility issues do not exist.

Nonetheless, it is reasonable to consider that end-users potentially have an expectation that traits must always be explicitly derived (which could lead to confusion).

#### Option 2

##### Unstable Rust
The [`StructuralEq`](https://doc.rust-lang.org/std/marker/trait.StructuralEq.html) marker indicates that constants of a type are suitable for use in patterns. It cannot be manually implemented; rather it is implied by `#[derive(Eq)]`. 

There are a few differences between this example and ours:
- `Eq` is not a formal supertrait of `StructuralEq`
- `StructuralEq` trait is not intended to ever be stabilized.

##### Third-Party Crates

###### Diesel
The Diesel ORM makes frequent use of this pattern. Nearly every `derive` it provides also derives other related traits. In some cases, these are super-traits. For instance, `derive(Identifiable)` [implies](https://github.com/diesel-rs/diesel/blob/a6b04d26f20c92e53ee273c04d809499b423494b/diesel_derives/src/identifiable.rs#L30-L60) a derivation of `HasTable`.

###### Bytemuck
Bytemuck's derives imply their supertraits. For example, `derive(AnyBitPattern)` implies `derive(Zeroable)`:
```rust
use bytemuck::{AnyBitPattern, Zeroable};

#[derive(AnyBitPattern, Copy, Clone)]
#[repr(C)]
struct Foo;

fn is_also_zeroable() -> impl Zeroable {
    Foo
}
```

##### Implied Traits at Use-Sites
Rust provides a similar ergonomic device at trait-use sites: Bringing a trait into scope implicitly brings the methods of its supertraits into scope. For example:
```rust
mod upstream {
    pub trait Super {
        fn sup(&self) {}
    }
    
    pub trait Sub: Super {}
    
    impl<T> Super for T {}
    impl<T> Sub for T {}
}

fn test<T>(t: T)
where
    T: upstream::Sub
{
    t.sup();
}
```
This compiles successfully.

## Conclusion

Option 2 offers a tighter development cycle in its pathological case than Option 1, and has precedent in the crates ecosystem.

--- a/src/byteorder.rs
+++ b/src/byteorder.rs
@@ -35,10 +35,10 @@
 //!
 //! ```rust,edition2021
 //! # #[cfg(feature = "derive")] { // This example uses derives, and won't compile without them
-//! use zerocopy::{IntoBytes, ByteSlice, FromBytes, FromZeros, NoCell, Ref, Unaligned};
+//! use zerocopy::{IntoBytes, ByteSlice, FromBytes, NoCell, Ref, Unaligned};
 //! use zerocopy::byteorder::network_endian::U16;
 //!
-//! #[derive(FromZeros, FromBytes, IntoBytes, NoCell, Unaligned)]
+//! #[derive(FromBytes, IntoBytes, NoCell, Unaligned)]
 //! #[repr(C)]
 //! struct UdpHeader {
 //!     src_port: U16,
@@ -357,7 +357,7 @@ example of how it can be used for parsing UDP packets.
 [`IntoBytes`]: crate::IntoBytes
 [`Unaligned`]: crate::Unaligned"),
             #[derive(Copy, Clone, Eq, PartialEq, Hash)]
-            #[cfg_attr(any(feature = "derive", test), derive(KnownLayout, NoCell, TryFromBytes, FromZeros, FromBytes, IntoBytes, Unaligned))]
+            #[cfg_attr(any(feature = "derive", test), derive(KnownLayout, NoCell, FromBytes, IntoBytes, Unaligned))]
             #[repr(transparent)]
             pub struct $name<O>([u8; $bytes], PhantomData<O>);
         }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -5766,9 +5766,7 @@ mod tests {
     //
     // This is used to test the custom derives of our traits. The `[u8]` type
     // gets a hand-rolled impl, so it doesn't exercise our custom derives.
-    #[derive(
-        Debug, Eq, PartialEq, TryFromBytes, FromZeros, FromBytes, IntoBytes, Unaligned, NoCell,
-    )]
+    #[derive(Debug, Eq, PartialEq, FromBytes, IntoBytes, Unaligned, NoCell)]
     #[repr(transparent)]
     struct Unsized([u8]);
 
@@ -7896,7 +7894,7 @@ mod tests {
             assert_eq!(too_many_bytes[0], 123);
         }
 
-        #[derive(Debug, Eq, PartialEq, TryFromBytes, FromZeros, FromBytes, IntoBytes, NoCell)]
+        #[derive(Debug, Eq, PartialEq, FromBytes, IntoBytes, NoCell)]
         #[repr(C)]
         struct Foo {
             a: u32,
@@ -7925,7 +7923,7 @@ mod tests {
 
     #[test]
     fn test_array() {
-        #[derive(TryFromBytes, FromZeros, FromBytes, IntoBytes, NoCell)]
+        #[derive(FromBytes, IntoBytes, NoCell)]
         #[repr(C)]
         struct Foo {
             a: [u16; 33],
@@ -7989,7 +7987,7 @@ mod tests {
 
     #[test]
     fn test_transparent_packed_generic_struct() {
-        #[derive(IntoBytes, TryFromBytes, FromZeros, FromBytes, Unaligned)]
+        #[derive(IntoBytes, FromBytes, Unaligned)]
         #[repr(transparent)]
         struct Foo<T> {
             _t: T,
@@ -7999,7 +7997,7 @@ mod tests {
         assert_impl_all!(Foo<u32>: FromZeros, FromBytes, IntoBytes);
         assert_impl_all!(Foo<u8>: Unaligned);
 
-        #[derive(IntoBytes, TryFromBytes, FromZeros, FromBytes, Unaligned)]
+        #[derive(IntoBytes, FromBytes, Unaligned)]
         #[repr(packed)]
         struct Bar<T, U> {
             _t: T,

--- a/src/util.rs
+++ b/src/util.rs
@@ -220,8 +220,6 @@ pub(crate) mod testutil {
     #[derive(
         KnownLayout,
         NoCell,
-        TryFromBytes,
-        FromZeros,
         FromBytes,
         IntoBytes,
         Eq,
@@ -249,9 +247,7 @@ pub(crate) mod testutil {
         }
     }
 
-    #[derive(
-        NoCell, FromZeros, FromBytes, Eq, PartialEq, Ord, PartialOrd, Default, Debug, Copy, Clone,
-    )]
+    #[derive(NoCell, FromBytes, Eq, PartialEq, Ord, PartialOrd, Default, Debug, Copy, Clone)]
     #[repr(C)]
     pub(crate) struct Nested<T, U: ?Sized> {
         _t: T,

--- a/src/wrappers.rs
+++ b/src/wrappers.rs
@@ -53,7 +53,7 @@ use super::*;
 #[derive(Default, Copy)]
 #[cfg_attr(
     any(feature = "derive", test),
-    derive(NoCell, KnownLayout, TryFromBytes, FromZeros, FromBytes, IntoBytes, Unaligned)
+    derive(NoCell, KnownLayout, FromBytes, IntoBytes, Unaligned)
 )]
 #[repr(C, packed)]
 pub struct Unalign<T>(T);

--- a/zerocopy-derive/src/lib.rs
+++ b/zerocopy-derive/src/lib.rs
@@ -295,13 +295,16 @@ pub fn derive_try_from_bytes(ts: proc_macro::TokenStream) -> proc_macro::TokenSt
 
 #[proc_macro_derive(FromZeros)]
 pub fn derive_from_zeros(ts: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let try_from_bytes = derive_try_from_bytes(ts.clone());
+
     let ast = syn::parse_macro_input!(ts as DeriveInput);
-    match &ast.data {
+    let from_zeros = match &ast.data {
         Data::Struct(strct) => derive_from_zeros_struct(&ast, strct),
         Data::Enum(enm) => derive_from_zeros_enum(&ast, enm),
         Data::Union(unn) => derive_from_zeros_union(&ast, unn),
     }
-    .into()
+    .into();
+    IntoIterator::into_iter([try_from_bytes, from_zeros]).collect()
 }
 
 /// Deprecated: prefer [`FromZeros`] instead.
@@ -314,13 +317,17 @@ pub fn derive_from_zeroes(ts: proc_macro::TokenStream) -> proc_macro::TokenStrea
 
 #[proc_macro_derive(FromBytes)]
 pub fn derive_from_bytes(ts: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let from_zeros = derive_from_zeros(ts.clone());
+
     let ast = syn::parse_macro_input!(ts as DeriveInput);
-    match &ast.data {
+    let from_bytes = match &ast.data {
         Data::Struct(strct) => derive_from_bytes_struct(&ast, strct),
         Data::Enum(enm) => derive_from_bytes_enum(&ast, enm),
         Data::Union(unn) => derive_from_bytes_union(&ast, unn),
     }
-    .into()
+    .into();
+
+    IntoIterator::into_iter([from_zeros, from_bytes]).collect()
 }
 
 #[proc_macro_derive(IntoBytes)]
@@ -447,25 +454,33 @@ fn derive_try_from_bytes_enum(ast: &DeriveInput, enm: &DataEnum) -> proc_macro2:
             .to_compile_error();
     }
 
-    // We don't actually care what the repr is; we just care that it's one of
-    // the allowed ones.
-    try_or_print!(ENUM_TRY_FROM_BYTES_CFG.validate_reprs(ast));
+    let reprs = try_or_print!(ENUM_TRY_FROM_BYTES_CFG.validate_reprs(ast));
+
+    // Figure out whether the enum could in theory implement `FromBytes`.
+    let from_bytes = enum_size_from_repr(reprs.as_slice())
+        .map(|size| {
+            // As of this writing, `enm.is_fieldless()` is redundant since we've
+            // already checked for it and returned if the check failed. However, if
+            // we ever remove that check, then without a similar check here, this
+            // code would become unsound.
+            enm.is_fieldless() && enm.variants.len() == 1usize << size
+        })
+        .unwrap_or(false);
+
     let variant_names = enm.variants.iter().map(|v| &v.ident);
-    let extras = Some(quote!(
-        // SAFETY: We use `is_bit_valid` to validate that the bit pattern
-        // corresponds to one of the field-less enum's variant discriminants.
-        // Thus, this is a sound implementation of `is_bit_valid`.
-        fn is_bit_valid(
-            candidate: ::zerocopy::Ptr<
-                '_,
-                Self,
-                (
-                    ::zerocopy::pointer::invariant::Shared,
-                    ::zerocopy::pointer::invariant::AnyAlignment,
-                    ::zerocopy::pointer::invariant::Initialized,
-                ),
-            >,
-        ) -> ::zerocopy::macro_util::core_reexport::primitive::bool {
+    let is_bit_valid_body = if from_bytes {
+        // If the enum could implement `FromBytes`, we can avoid emitting a
+        // match statement. This is faster to compile, and generates code which
+        // performs better.
+        quote!({
+            // Prevent an "unused" warning.
+            let _ = candidate;
+            // SAFETY: If the enum could implement `FromBytes`, then all bit
+            // patterns are valid. Thus, this is a sound implementation.
+            true
+        })
+    } else {
+        quote!(
             use ::zerocopy::macro_util::core_reexport;
             // SAFETY:
             // - `cast` is implemented as required.
@@ -499,6 +514,25 @@ fn derive_try_from_bytes_enum(ast: &DeriveInput, enm: &DataEnum) -> proc_macro2:
                 // `candidate` refers to a bit-valid `Self`.
                 discriminant == d
             })*
+        )
+    };
+
+    let extras = Some(quote!(
+        // SAFETY: We use `is_bit_valid` to validate that the bit pattern
+        // corresponds to one of the field-less enum's variant discriminants.
+        // Thus, this is a sound implementation of `is_bit_valid`.
+        fn is_bit_valid(
+            candidate: ::zerocopy::Ptr<
+                '_,
+                Self,
+                (
+                    ::zerocopy::pointer::invariant::Shared,
+                    ::zerocopy::pointer::invariant::AnyAlignment,
+                    ::zerocopy::pointer::invariant::Initialized,
+                ),
+            >,
+        ) -> ::zerocopy::macro_util::core_reexport::primitive::bool {
+            #is_bit_valid_body
         }
     ));
     impl_block(ast, enm, Trait::TryFromBytes, FieldBounds::ALL_SELF, SelfBounds::None, None, extras)
@@ -608,13 +642,9 @@ fn derive_from_bytes_enum(ast: &DeriveInput, enm: &DataEnum) -> proc_macro2::Tok
 
     let reprs = try_or_print!(ENUM_FROM_BYTES_CFG.validate_reprs(ast));
 
-    let variants_required = match reprs.as_slice() {
-        [EnumRepr::U8] | [EnumRepr::I8] => 1usize << 8,
-        [EnumRepr::U16] | [EnumRepr::I16] => 1usize << 16,
-        // `validate_reprs` has already validated that it's one of the preceding
-        // patterns.
-        _ => unreachable!(),
-    };
+    let variants_required = 1usize
+        << enum_size_from_repr(reprs.as_slice())
+            .expect("internal error: `validate_reprs` has already validated that the reprs guarantee the enum's size");
     if enm.variants.len() != variants_required {
         return Error::new_spanned(
             ast,
@@ -627,6 +657,15 @@ fn derive_from_bytes_enum(ast: &DeriveInput, enm: &DataEnum) -> proc_macro2::Tok
     }
 
     impl_block(ast, enm, Trait::FromBytes, FieldBounds::ALL_SELF, SelfBounds::None, None, None)
+}
+
+// Returns `None` if the enum's size is not guaranteed by the repr.
+fn enum_size_from_repr(reprs: &[EnumRepr]) -> Option<usize> {
+    match reprs {
+        [EnumRepr::U8] | [EnumRepr::I8] => Some(8),
+        [EnumRepr::U16] | [EnumRepr::I16] => Some(16),
+        _ => None,
+    }
 }
 
 #[rustfmt::skip]

--- a/zerocopy-derive/tests/enum_from_bytes.rs
+++ b/zerocopy-derive/tests/enum_from_bytes.rs
@@ -29,7 +29,7 @@ include!("include.rs");
 // `Variant128` has a discriminant of -128) since Rust won't automatically wrap
 // a signed discriminant around without you explicitly telling it to.
 
-#[derive(imp::FromZeros, imp::FromBytes)]
+#[derive(imp::FromBytes)]
 #[repr(u8)]
 enum FooU8 {
     Variant0,
@@ -292,7 +292,7 @@ enum FooU8 {
 
 util_assert_impl_all!(FooU8: imp::FromBytes);
 
-#[derive(imp::FromZeros, imp::FromBytes)]
+#[derive(imp::FromBytes)]
 #[repr(i8)]
 enum FooI8 {
     Variant0,
@@ -555,7 +555,7 @@ enum FooI8 {
 
 util_assert_impl_all!(FooI8: imp::FromBytes);
 
-#[derive(imp::FromZeros, imp::FromBytes)]
+#[derive(imp::FromBytes)]
 #[repr(u8, align(2))]
 enum FooU8Align {
     Variant0,
@@ -818,7 +818,7 @@ enum FooU8Align {
 
 util_assert_impl_all!(FooU8Align: imp::FromBytes);
 
-#[derive(imp::FromZeros, imp::FromBytes)]
+#[derive(imp::FromBytes)]
 #[repr(i8, align(2))]
 enum FooI8Align {
     Variant0,
@@ -1081,7 +1081,7 @@ enum FooI8Align {
 
 util_assert_impl_all!(FooI8Align: imp::FromBytes);
 
-#[derive(imp::FromZeros, imp::FromBytes)]
+#[derive(imp::FromBytes)]
 #[repr(u16)]
 enum FooU16 {
     Variant0,
@@ -66624,7 +66624,7 @@ enum FooU16 {
 
 util_assert_impl_all!(FooU16: imp::FromBytes);
 
-#[derive(imp::FromZeros, imp::FromBytes)]
+#[derive(imp::FromBytes)]
 #[repr(i16)]
 enum FooI16 {
     Variant0,

--- a/zerocopy-derive/tests/hygiene.rs
+++ b/zerocopy-derive/tests/hygiene.rs
@@ -18,14 +18,7 @@ include!("include.rs");
 
 extern crate zerocopy as _zerocopy;
 
-// #[macro_use]
-// mod util;
-
-// use std::{marker::PhantomData, option::IntoIter};
-
-#[derive(
-    _zerocopy::KnownLayout, _zerocopy::FromZeros, _zerocopy::FromBytes, _zerocopy::Unaligned,
-)]
+#[derive(_zerocopy::KnownLayout, _zerocopy::FromBytes, _zerocopy::Unaligned)]
 #[repr(C)]
 struct TypeParams<'a, T, I: imp::Iterator> {
     a: T,

--- a/zerocopy-derive/tests/include.rs
+++ b/zerocopy-derive/tests/include.rs
@@ -47,8 +47,6 @@ pub mod util {
     #[derive(
         super::imp::KnownLayout,
         super::imp::NoCell,
-        super::imp::TryFromBytes,
-        super::imp::FromZeros,
         super::imp::FromBytes,
         super::imp::IntoBytes,
         Copy,

--- a/zerocopy-derive/tests/paths_and_modules.rs
+++ b/zerocopy-derive/tests/paths_and_modules.rs
@@ -17,13 +17,13 @@ include!("include.rs");
 mod foo {
     use super::*;
 
-    #[derive(imp::FromZeros, imp::FromBytes, imp::IntoBytes, imp::Unaligned)]
+    #[derive(imp::FromBytes, imp::IntoBytes, imp::Unaligned)]
     #[repr(C)]
     pub struct Foo {
         foo: u8,
     }
 
-    #[derive(imp::FromZeros, imp::FromBytes, imp::IntoBytes, imp::Unaligned)]
+    #[derive(imp::FromBytes, imp::IntoBytes, imp::Unaligned)]
     #[repr(C)]
     pub struct Bar {
         bar: u8,
@@ -32,7 +32,7 @@ mod foo {
 
 use foo::Foo;
 
-#[derive(imp::FromZeros, imp::FromBytes, imp::IntoBytes, imp::Unaligned)]
+#[derive(imp::FromBytes, imp::IntoBytes, imp::Unaligned)]
 #[repr(C)]
 struct Baz {
     foo: Foo,

--- a/zerocopy-derive/tests/struct_from_bytes.rs
+++ b/zerocopy-derive/tests/struct_from_bytes.rs
@@ -15,19 +15,19 @@ include!("include.rs");
 // A struct is `FromBytes` if:
 // - all fields are `FromBytes`
 
-#[derive(imp::FromZeros, imp::FromBytes)]
+#[derive(imp::FromBytes)]
 struct Zst;
 
 util_assert_impl_all!(Zst: imp::FromBytes);
 
-#[derive(imp::FromZeros, imp::FromBytes)]
+#[derive(imp::FromBytes)]
 struct One {
     a: u8,
 }
 
 util_assert_impl_all!(One: imp::FromBytes);
 
-#[derive(imp::FromZeros, imp::FromBytes)]
+#[derive(imp::FromBytes)]
 struct Two {
     a: u8,
     b: Zst,
@@ -35,14 +35,14 @@ struct Two {
 
 util_assert_impl_all!(Two: imp::FromBytes);
 
-#[derive(imp::FromZeros, imp::FromBytes)]
+#[derive(imp::FromBytes)]
 struct Unsized {
     a: [u8],
 }
 
 util_assert_impl_all!(Unsized: imp::FromBytes);
 
-#[derive(imp::FromZeros, imp::FromBytes)]
+#[derive(imp::FromBytes)]
 struct TypeParams<'a, T: ?imp::Sized, I: imp::Iterator> {
     a: I::Item,
     b: u8,
@@ -58,7 +58,7 @@ util_assert_impl_all!(TypeParams<'static, [util::AU16], imp::IntoIter<()>>: imp:
 
 // Deriving `FromBytes` should work if the struct has bounded parameters.
 
-#[derive(imp::FromZeros, imp::FromBytes)]
+#[derive(imp::FromBytes)]
 #[repr(transparent)]
 struct WithParams<'a: 'b, 'b: 'a, T: 'a + 'b + imp::FromBytes, const N: usize>(
     [T; N],

--- a/zerocopy-derive/tests/struct_try_from_bytes.rs
+++ b/zerocopy-derive/tests/struct_try_from_bytes.rs
@@ -26,7 +26,7 @@ fn zst() {
     imp::assert!(is_bit_valid);
 }
 
-#[derive(imp::TryFromBytes, imp::FromZeros, imp::FromBytes)]
+#[derive(imp::FromBytes)]
 #[repr(C)]
 struct One {
     a: u8,
@@ -45,7 +45,7 @@ fn one() {
     imp::assert!(is_bit_valid);
 }
 
-#[derive(imp::TryFromBytes, imp::FromZeros)]
+#[derive(imp::FromZeros)]
 #[repr(C)]
 struct Two {
     a: bool,
@@ -88,7 +88,7 @@ fn two_bad() {
     imp::assert!(!is_bit_valid);
 }
 
-#[derive(imp::TryFromBytes, imp::FromZeros, imp::FromBytes)]
+#[derive(imp::FromBytes)]
 #[repr(C)]
 struct Unsized {
     a: [u8],
@@ -119,7 +119,7 @@ fn un_sized() {
     imp::assert!(is_bit_valid);
 }
 
-#[derive(imp::TryFromBytes, imp::FromZeros, imp::FromBytes)]
+#[derive(imp::FromBytes)]
 #[repr(C)]
 struct TypeParams<'a, T: ?imp::Sized, I: imp::Iterator> {
     a: I::Item,
@@ -136,7 +136,7 @@ util_assert_impl_all!(TypeParams<'static, [util::AU16], imp::IntoIter<()>>: imp:
 
 // Deriving `imp::TryFromBytes` should work if the struct has bounded parameters.
 
-#[derive(imp::TryFromBytes, imp::FromZeros, imp::FromBytes)]
+#[derive(imp::FromBytes)]
 #[repr(transparent)]
 struct WithParams<'a: 'b, 'b: 'a, T: 'a + 'b + imp::TryFromBytes, const N: usize>(
     imp::PhantomData<&'a &'b ()>,

--- a/zerocopy-derive/tests/union_from_bytes.rs
+++ b/zerocopy-derive/tests/union_from_bytes.rs
@@ -15,21 +15,21 @@ include!("include.rs");
 // A union is `imp::FromBytes` if:
 // - all fields are `imp::FromBytes`
 
-#[derive(Clone, Copy, imp::FromZeros, imp::FromBytes)]
+#[derive(Clone, Copy, imp::NoCell, imp::FromBytes)]
 union Zst {
     a: (),
 }
 
 util_assert_impl_all!(Zst: imp::FromBytes);
 
-#[derive(imp::FromZeros, imp::FromBytes)]
+#[derive(imp::NoCell, imp::FromBytes)]
 union One {
     a: u8,
 }
 
 util_assert_impl_all!(One: imp::FromBytes);
 
-#[derive(imp::FromZeros, imp::FromBytes)]
+#[derive(imp::NoCell, imp::FromBytes)]
 union Two {
     a: u8,
     b: Zst,
@@ -37,7 +37,7 @@ union Two {
 
 util_assert_impl_all!(Two: imp::FromBytes);
 
-#[derive(imp::FromZeros, imp::FromBytes)]
+#[derive(imp::NoCell, imp::FromBytes)]
 union TypeParams<'a, T: imp::Copy, I: imp::Iterator>
 where
     I::Item: imp::Copy,
@@ -54,7 +54,7 @@ util_assert_impl_all!(TypeParams<'static, (), imp::IntoIter<()>>: imp::FromBytes
 
 // Deriving `imp::FromBytes` should work if the union has bounded parameters.
 
-#[derive(imp::FromZeros, imp::FromBytes)]
+#[derive(imp::NoCell, imp::FromBytes)]
 #[repr(C)]
 union WithParams<'a: 'b, 'b: 'a, T: 'a + 'b + imp::FromBytes, const N: usize>
 where

--- a/zerocopy-derive/tests/union_from_zeros.rs
+++ b/zerocopy-derive/tests/union_from_zeros.rs
@@ -15,21 +15,21 @@ include!("include.rs");
 // A union is `imp::FromZeros` if:
 // - all fields are `imp::FromZeros`
 
-#[derive(Clone, Copy, imp::FromZeros)]
+#[derive(Clone, Copy, imp::NoCell, imp::FromZeros)]
 union Zst {
     a: (),
 }
 
 util_assert_impl_all!(Zst: imp::FromZeros);
 
-#[derive(imp::FromZeros)]
+#[derive(imp::NoCell, imp::FromZeros)]
 union One {
     a: bool,
 }
 
 util_assert_impl_all!(One: imp::FromZeros);
 
-#[derive(imp::FromZeros)]
+#[derive(imp::NoCell, imp::FromZeros)]
 union Two {
     a: bool,
     b: Zst,
@@ -37,7 +37,7 @@ union Two {
 
 util_assert_impl_all!(Two: imp::FromZeros);
 
-#[derive(imp::FromZeros)]
+#[derive(imp::NoCell, imp::FromZeros)]
 union TypeParams<'a, T: imp::Copy, I: imp::Iterator>
 where
     I::Item: imp::Copy,
@@ -54,7 +54,7 @@ util_assert_impl_all!(TypeParams<'static, (), imp::IntoIter<()>>: imp::FromZeros
 
 // Deriving `imp::FromZeros` should work if the union has bounded parameters.
 
-#[derive(imp::FromZeros)]
+#[derive(imp::NoCell, imp::FromZeros)]
 #[repr(C)]
 union WithParams<'a: 'b, 'b: 'a, T: 'a + 'b + imp::FromZeros, const N: usize>
 where

--- a/zerocopy-derive/tests/union_try_from_bytes.rs
+++ b/zerocopy-derive/tests/union_try_from_bytes.rs
@@ -15,7 +15,7 @@ include!("include.rs");
 // A struct is `imp::TryFromBytes` if:
 // - any of its fields are `imp::TryFromBytes`
 
-#[derive(imp::NoCell, imp::TryFromBytes, imp::FromZeros, imp::FromBytes)]
+#[derive(imp::NoCell, imp::FromBytes)]
 union One {
     a: u8,
 }
@@ -33,7 +33,7 @@ fn one() {
     assert!(is_bit_valid);
 }
 
-#[derive(imp::NoCell, imp::TryFromBytes, imp::FromZeros)]
+#[derive(imp::NoCell, imp::FromZeros)]
 #[repr(C)]
 union Two {
     a: bool,
@@ -83,7 +83,7 @@ fn two_bad() {
     assert!(!is_bit_valid);
 }
 
-#[derive(imp::NoCell, imp::TryFromBytes, imp::FromZeros)]
+#[derive(imp::NoCell, imp::FromZeros)]
 #[repr(C)]
 union BoolAndZst {
     a: bool,
@@ -113,7 +113,7 @@ fn bool_and_zst() {
     assert!(is_bit_valid);
 }
 
-#[derive(imp::NoCell, imp::TryFromBytes, imp::FromZeros, imp::FromBytes)]
+#[derive(imp::NoCell, imp::FromBytes)]
 #[repr(C)]
 union TypeParams<'a, T: imp::Copy, I: imp::Iterator>
 where
@@ -133,7 +133,7 @@ util_assert_impl_all!(TypeParams<'static, [util::AU16; 2], imp::IntoIter<()>>: i
 
 // Deriving `imp::TryFromBytes` should work if the union has bounded parameters.
 
-#[derive(imp::NoCell, imp::TryFromBytes, imp::FromZeros, imp::FromBytes)]
+#[derive(imp::NoCell, imp::FromBytes)]
 #[repr(C)]
 union WithParams<'a: 'b, 'b: 'a, T: 'a + 'b + imp::TryFromBytes, const N: usize>
 where

